File: web/js/views/tokens.js
// Token efficiency: where prompt and output tokens go, cache hits, trend and the settings that send less
// (orchestrator/tokens.py, docs/TOKEN_EFFICIENCY.md). Self-contained: GET /api/tokens.
import { $, $$, esc, icon, toast } from "../ui.js";
import { get, api } from "../api.js";
import { agentLabel } from "../state.js";

const SECTION_LABEL = {
  rules: "Rules", protocol: "Session protocol", tools: "Tools", task: "Task", lessons: "Lessons", context: "Repositories & system map",
  design: "Design", environment: "Environment", packet: "Context packet / plan", judge: "Acceptance contract", report: "Worker report",
  verification: "Verification", diff: "Diff", guidance: "Guidance", instruction: "Instructions", transcript: "Replayed transcript", other: "Other",
};
const n = (v) => { v = Number(v) || 0; return v >= 1e6 ? `${(v / 1e6).toFixed(2)}M` : v >= 1e4 ? `${Math.round(v / 1e3)}k` : v >= 1e3 ? `${(v / 1e3).toFixed(1)}k` : String(v); };
const pct = (v) => `${Math.round((Number(v) || 0) * 100)}%`;
const usd = (v) => `$${(Number(v) || 0).toFixed((Number(v) || 0) < 1 ? 3 : 2)}`;

function bars(rows, { value, label, title, max }) {
  const top = max || Math.max(1, ...rows.map(value));
  return `<div class="tk-bars">${rows.map((r) => {
    const v = value(r);
    return `<div class="tk-bar" title="${esc(title(r))}"><span class="tk-bar-label truncate">${esc(label(r))}</span><span class="tk-track"><span class="tk-fill" style="width:${Math.max(0.5, (v / top) * 100).toFixed(1)}%"></span></span><span class="tk-bar-val">${esc(n(v))}</span></div>`;
  }).join("")}</div>`;
}

function table(rows, head) {
  return `<div class="tk-table-wrap"><table class="tk-table"><thead><tr>${head.map(([h, cls]) => `<th class="${cls || ""}">${esc(h)}</th>`).join("")}</tr></thead><tbody>${rows.join("")}</tbody></table></div>`;
}

export function mountTokens(body) {
  let days = 30;
  try { days = Number(localStorage.getItem("relay.tokenDays")) || 30; } catch {}
  const load = async () => {
    let d;
    try { d = await get(`/api/tokens?days=${days}`); } catch (e) { body.innerHTML = `<div class="card"><div class="card-body"><div class="empty small">${icon("alert")} ${esc(e.message)}</div></div></div>`; return; }
    draw(d);
  };
  const draw = (d) => {
    const t = d.total, s = d.settings || {};
    const secs = Object.entries(d.sections || {});
    const secTotal = secs.reduce((a, [, v]) => a + v.tokens, 0);
    const roleRows = Object.entries(d.roles || {}).map(([k, v]) => `<tr><td>${esc(k)}</td><td class="num">${v.turns}</td><td class="num">${n(v.prompt)}</td><td class="num">${pct(v.cache_ratio)}</td><td class="num">${n(v.output)}</td><td class="num">${n(v.turns ? v.prompt / v.turns : 0)}</td><td class="num">${usd(v.cost_usd)}</td></tr>`);
    const agentRows = Object.entries(d.agents || {}).map(([k, v]) => `<tr><td>${esc(agentLabel(k))}</td><td class="num">${v.turns}</td><td class="num">${n(v.prompt)}</td><td class="num">${pct(v.cache_ratio)}</td><td class="num">${n(v.output)}</td><td class="num">${n(v.turns ? v.prompt / v.turns : 0)}</td><td class="num">${usd(v.cost_usd)}</td></tr>`);
    const head = [["", ""], ["Turns", "num"], ["Prompt tokens", "num"], ["Cached", "num"], ["Output", "num"], ["Prompt / turn", "num"], ["Cost", "num"]];
    body.innerHTML = `
      <div class="card"><div class="card-head"><div><h3>Token efficiency</h3><p class="card-sub">Where the tokens of agent turns go, from what each CLI reported. Prompt tokens include cache reads; cached tokens cost about a tenth.</p></div>
        <select id="tkDays" aria-label="Period">${[7, 30, 90].map((x) => `<option value="${x}" ${x === days ? "selected" : ""}>Last ${x} days</option>`).join("")}</select></div>
        <div class="card-body">
          <div class="tk-kpis">
            <div class="tk-kpi"><div class="lbl">Prompt tokens</div><div class="val">${n(t.prompt)}</div><div class="sub">${t.turns} turns</div></div>
            <div class="tk-kpi"><div class="lbl">Cache hits</div><div class="val">${pct(t.cache_ratio)}</div><div class="sub">${n(t.cached)} cached</div></div>
            <div class="tk-kpi"><div class="lbl">Output tokens</div><div class="val">${n(t.output)}</div><div class="sub">${pct(t.output_share)} of all tokens</div></div>
            <div class="tk-kpi"><div class="lbl">Cost</div><div class="val">${usd(t.cost_usd)}</div><div class="sub">${t.turns ? usd(t.cost_usd / t.turns) : "$0"} per turn</div></div>
          </div>
        </div></div>
      <div class="card"><div class="card-head"><div><h3>What the prompts are made of</h3><p class="card-sub">Relay's own part of each prompt by section (characters ÷ 4), over ${d.measured_turns} measured turn${d.measured_turns === 1 ? "" : "s"}. The CLI's system prompt, tool definitions and the tool results inside a turn come on top.</p></div></div>
        <div class="card-body">${secs.length ? bars(secs, { value: ([, v]) => v.tokens, label: ([k]) => SECTION_LABEL[k] || k, title: ([k, v]) => `${SECTION_LABEL[k] || k}: ≈${v.tokens.toLocaleString()} tokens (${secTotal ? Math.round((v.tokens / secTotal) * 100) : 0}%)` })
          : `<div class="empty small">${icon("gauge", "lg")}<p>No measured turns yet. Every agent turn from now on records its prompt sections.</p></div>`}</div></div>
      <div class="card"><div class="card-head"><h3>By role</h3></div><div class="card-body">${roleRows.length ? table(roleRows, head) : '<div class="muted">No turns in this period.</div>'}</div></div>
      <div class="card"><div class="card-head"><h3>By agent</h3></div><div class="card-body">${agentRows.length ? table(agentRows, head) : '<div class="muted">No turns in this period.</div>'}</div></div>
      <div class="card"><div class="card-head"><div><h3>Trend</h3><p class="card-sub">Prompt tokens per day; hover a bar for cached, output and cost.</p></div></div><div class="card-body">${d.trend.length ? bars(d.trend, { value: (r) => r.prompt, label: (r) => r.day, title: (r) => `${r.day}: ${r.prompt.toLocaleString()} prompt · ${pct(r.cache_ratio)} cached · ${r.output.toLocaleString()} output · ${usd(r.cost_usd)} · ${r.turns} turns` }) : '<div class="muted">No turns in this period.</div>'}</div></div>
      <div class="card"><div class="card-head"><h3>Most expensive tasks</h3></div><div class="card-body">${d.tasks.length ? table(d.tasks.map((r) => `<tr><td class="tk-name"><a href="#/task/${esc(r.id)}">${r.number ? `#${esc(r.number)} ` : ""}${esc(r.name || r.id)}</a></td><td class="num">${r.turns}</td><td class="num">${n(r.prompt)}</td><td class="num">${pct(r.cache_ratio)}</td><td class="num">${n(r.output)}</td><td class="num">${n(r.turns ? r.prompt / r.turns : 0)}</td><td class="num">${usd(r.cost_usd)}</td></tr>`), [["Task", ""], ...head.slice(1)]) : '<div class="muted">No tasks in this period.</div>'}</div></div>
      <div class="card"><div class="card-head"><div><h3>Send less</h3><p class="card-sub">Applied to every new agent turn. See docs/TOKEN_EFFICIENCY.md for what each one does and what it saved.</p></div></div><div class="card-body">
        <div class="field inline"><label>Lean agent context (Claude)<div class="help">Agents run without your personal plugins, skills, slash commands and claude.ai connectors; the task's tools come from Relay. Measured: about 6k fewer tokens on every API call.</div></label><span class="switch ${s.token_lean_agent_context !== false ? "on" : ""}" data-tk="token_lean_agent_context" role="switch" tabindex="0"></span></div>
        <div class="field inline"><label>Do not re-send what a live session already has<div class="help">Unchanged acceptance contract and verification results become one line; the reply format is explained once.</div></label><span class="switch ${s.token_prompt_deltas !== false ? "on" : ""}" data-tk="token_prompt_deltas" role="switch" tabindex="0"></span></div>
        <div class="field inline"><label>Failing checks: send the lines that explain the failure<div class="help">The runner's failure summary and error lines, deduplicated, instead of the tail of the log.</div></label><span class="switch ${s.token_failure_excerpts !== false ? "on" : ""}" data-tk="token_failure_excerpts" role="switch" tabindex="0"></span></div>
        <div class="grid3">
          <div class="field"><label for="tkDiff">Diff for the reviewer</label><select id="tkDiff" data-tkv="token_diff_mode"><option value="targeted" ${s.token_diff_mode !== "full" ? "selected" : ""}>Targeted: stat, small files, git command for the rest</option><option value="full" ${s.token_diff_mode === "full" ? "selected" : ""}>Full diff up to the budget</option></select></div>
          <div class="field"><label for="tkCompact">Compact a session above (tokens)</label><input id="tkCompact" type="number" min="0" step="10000" data-tkv="token_compact_threshold" value="${esc(s.token_compact_threshold ?? 120000)}"><div class="help">A fresh session starts from Relay's handoff. 0 turns it off.</div></div>
          <div class="field"><label for="tkVerb">Codex verbosity</label><select id="tkVerb" data-tkv="token_codex_verbosity">${["", "low", "medium", "high"].map((v) => `<option value="${v}" ${(s.token_codex_verbosity || "") === v ? "selected" : ""}>${v || "CLI default"}</option>`).join("")}</select></div>
          <div class="field"><label for="tkClaudeOut">Claude max output per response</label><input id="tkClaudeOut" type="number" min="0" step="1000" data-tkv="token_claude_max_output_tokens" value="${esc(s.token_claude_max_output_tokens || 0)}"><div class="help">0 keeps Claude Code's default. Too low truncates large file writes.</div></div>
        </div>
      </div></div>`;
    $("#tkDays", body).onchange = (e) => { days = Number(e.target.value); try { localStorage.setItem("relay.tokenDays", String(days)); } catch {} load(); };
    const save = async (patch) => { try { await api.saveSettings(patch); toast("success", "Saved"); } catch (e) { toast("error", "Could not save", e.message); } };
    $$("[data-tk]", body).forEach((sw) => {
      const flip = () => { sw.classList.toggle("on"); save({ [sw.dataset.tk]: sw.classList.contains("on") }); };
      sw.onclick = flip; sw.onkeydown = (e) => { if (e.key === " " || e.key === "Enter") { e.preventDefault(); flip(); } };
    });
    $$("[data-tkv]", body).forEach((i) => (i.onchange = () => save({ [i.dataset.tkv]: i.type === "number" ? Number(i.value) || 0 : i.value })));
  };
  body.innerHTML = `<div class="card"><div class="card-body"><div class="empty small">${icon("spinner", "spin")} Loading…</div></div></div>`;
  load();
}

File: orchestrator/tokens.py
"""Token efficiency: measure where prompt tokens go, and the helpers that send less.

Measuring
  Prompt        a str that also carries its sections ({name: chars}) and where context blocks are inserted.
                protocol.py builds every prompt from named parts, so each turn records how much the rules,
                task, context packet, diff, verification, lessons, design and tools cost (chars / 4 ≈ tokens).
  report()      aggregates per role, agent, section and day across tasks for the Token efficiency panel.

Sending less
  failure_excerpt()  the lines of a failing check that explain it (pytest/jest/go/cargo summaries, errors),
                     deduplicated, instead of the tail of the whole log.
  targeted_diff()    diffstat plus whole hunks for small files within the budget; larger files are listed
                     with the command that shows them. On a re-review only files that changed since the
                     last round are sent.
  compact_handoff()  the state a fresh session needs when a long session is restarted to shed context.
"""
from __future__ import annotations

import hashlib
import re
import time
from collections import OrderedDict
from datetime import datetime, timezone

CHARS_PER_TOKEN = 4

# Sections in the order they appear in a kickoff prompt: stable first (prefix-cache friendly), volatile last.
SECTION_ORDER = ["rules", "protocol", "tools", "task", "lessons", "knowledge", "context", "design", "environment", "packet",
                 "judge", "report", "verification", "diff", "guidance", "instruction", "other"]


class Prompt(str):
    """A prompt string that remembers the size of its named sections."""
    sections: dict
    insert_at: int
    full: str | None

    def __new__(cls, text: str, sections: dict | None = None, insert_at: int | None = None, full: str | None = None):
        obj = super().__new__(cls, text)
        obj.sections = dict(sections or {})
        obj.insert_at = len(text) if insert_at is None else insert_at
        obj.full = full
        return obj


def build(parts, insert_before: str | None = None, full: str | None = None) -> Prompt:
    """Join (section, text) parts with blank lines; empty parts are dropped.

    insert_before names the first section that context blocks (lessons, system map) are placed in front of.
    """
    chunks, sections, insert_at = [], OrderedDict(), None
    pos = 0
    for name, text in parts:
        text = (text or "").strip("\n")
        if not text.strip():
            continue
        if insert_before and name == insert_before and insert_at is None:
            insert_at = pos
        chunks.append(text)
        sections[name] = sections.get(name, 0) + len(text) + 2
        pos += len(text) + 2
    body = "\n\n".join(chunks) + "\n"
    return Prompt(body, sections, insert_at, full)


def insert_block(prompt: str, block: str, section: str) -> str:
    """Insert a context block where the prompt wants it (before its volatile tail), keeping the accounting."""
    if not block or not block.strip():
        return prompt
    block = block.strip("\n")
    if isinstance(prompt, Prompt):
        at = min(prompt.insert_at, len(prompt))
        text = prompt[:at] + block + "\n\n" + prompt[at:]
        sections = dict(prompt.sections)
        sections[section] = sections.get(section, 0) + len(block) + 2
        full = insert_block(prompt.full, block, section) if prompt.full else None
        return Prompt(text, sections, at + len(block) + 2, full)
    at = prompt.find("HOW THIS SESSION WORKS")
    if at < 0:
        return prompt.rstrip() + "\n\n" + block + "\n"
    return prompt[:at] + block + "\n\n" + prompt[at:]


def prepend(prompt: str, text: str, section: str = "other") -> str:
    if not text:
        return prompt
    sections = dict(getattr(prompt, "sections", None) or {"other": len(prompt)})
    sections[section] = sections.get(section, 0) + len(text) + 2
    full = getattr(prompt, "full", None)
    return Prompt(text + "\n\n" + prompt, sections, None, (text + "\n\n" + full) if full else None)


def sections_of(prompt: str) -> dict:
    s = getattr(prompt, "sections", None)
    if s:
        return dict(s)
    return {"other": len(prompt or "")}


def est(chars: int) -> int:
    return int(round((chars or 0) / CHARS_PER_TOKEN))


def digest(text: str) -> str:
    return hashlib.sha1((text or "").encode("utf-8", "replace")).hexdigest()[:16]


# ============================================================================ usage normalization
# CLIs whose reported input tokens exclude cache reads (Anthropic-style usage); for the others input includes them.
INPUT_EXCLUDES_CACHED = {"claude", "qwen", "amp", "cursor", "opencode", "kilo", "crush"}


def prompt_tokens(agent: str, usage: dict) -> int:
    inp, cached = int(usage.get("input") or 0), int(usage.get("cached") or 0)
    write = int(usage.get("cache_write") or 0)
    return inp + cached + write if agent in INPUT_EXCLUDES_CACHED else max(inp, cached)


def cache_ratio(agent: str, usage: dict) -> float:
    total = prompt_tokens(agent, usage)
    return round(min(1.0, int(usage.get("cached") or 0) / total), 3) if total else 0.0


# ============================================================================ verification
_FAIL_LINE = re.compile(r"(?i)(^FAILED |^ERROR |^E\s{2,}|error[:\[]|exception|traceback|assert|\bfail(?:ed|ure)?\b|✕|●|panic:|"
                        r"cannot find|not found|undefined|expected|received|^\s*at .*:\d+|^--- FAIL|:\d+:\d+: error)")
_NOISE = re.compile(r"^\s*$|^[=\-_.·]{8,}\s*$|^\s*\d+%\||DeprecationWarning|^npm (?:notice|WARN)")
_SUMMARY_START = re.compile(r"(?i)^=+ (?:short test summary info|FAILURES|ERRORS) =+$|^Summary of all failing tests|^failures:$")


def failure_excerpt(output: str, limit: int) -> str:
    """The part of a failing check's output that explains the failure, within `limit` characters.

    Order of preference: a test runner's own failure summary, then error lines with a little context,
    then the tail. Repeated lines (the same stack frame, the same warning) are kept once.
    """
    text = (output or "").replace("\r", "")
    if len(text) <= limit:
        return text
    lines = text.splitlines()
    picked: list[str] = []
    for i, line in enumerate(lines):
        if _SUMMARY_START.match(line.strip()):
            picked = lines[i:]
            break
    if not picked:
        keep = set()
        for i, line in enumerate(lines):
            if _FAIL_LINE.search(line) and not _NOISE.search(line):
                keep.update(range(max(0, i - 1), min(len(lines), i + 3)))
        picked = [lines[i] for i in sorted(keep)]
    seen, out = set(), []
    for line in picked:
        key = re.sub(r"\d+", "#", line.strip())
        if _NOISE.search(line) or key in seen:
            continue
        seen.add(key)
        out.append(line)
    tail = [x for x in lines[-6:] if x.strip()]  # the runner's final verdict line ("3 failed, 40 passed")
    body = "\n".join(out)
    tail_text = "\n".join(x for x in tail if x not in out)
    room = max(200, limit - len(tail_text) - 40)
    if len(body) > room:
        body = body[:room].rsplit("\n", 1)[0] + f"\n… ({len(out)} relevant lines, trimmed)"
    result = (body + ("\n…\n" + tail_text if tail_text else "")).strip()
    return result[:limit] if result else text[-limit:]


# ============================================================================ diffs
_FILE_HEAD = re.compile(r"^diff --git a/(.+?) b/(.+)$", re.M)


def split_diff(diff: str) -> list[tuple[str, str]]:
    starts = [(m.start(), m.group(2)) for m in _FILE_HEAD.finditer(diff or "")]
    out = []
    for i, (pos, path) in enumerate(starts):
        end = starts[i + 1][0] if i + 1 < len(starts) else len(diff)
        out.append((path, diff[pos:end]))
    if not starts and (diff or "").strip():
        out.append(("(diff)", diff))
    return out


def targeted_diff(diff: str, budget: int, base: str = "", previous: dict | None = None) -> tuple[str, dict]:
    """Stat + whole hunks for as many files as fit, smallest first; the rest listed with the command to see them.

    previous: {path: digest} from the last round; files whose diff did not change since are listed, not re-sent.
    Returns (text, {path: digest}).
    """
    untracked = []
    for m in re.finditer(r"\n*UNTRACKED FILES:\n(.*?)(?=\n### |\Z)", diff or "", re.S):
        untracked += [x.strip() for x in m.group(1).splitlines() if x.strip()]
    diff = re.sub(r"\n*UNTRACKED FILES:\n.*?(?=\n### |\Z)", "", diff or "", flags=re.S)
    files = split_diff(diff)
    hashes = {p: digest(body) for p, body in files}
    if untracked:
        hashes["(untracked)"] = digest("\n".join(untracked))
    if not files:
        return ("UNTRACKED FILES (new, not in git diff; read them directly):\n" + "\n".join(untracked)) if untracked else "(no changes)", hashes
    rows, send, unchanged = [], [], []
    for path, body in files:
        adds = sum(1 for x in body.splitlines() if x.startswith("+") and not x.startswith("+++"))
        dels = sum(1 for x in body.splitlines() if x.startswith("-") and not x.startswith("---"))
        rows.append((path, adds, dels, body))
    stat = "\n".join(f"  {p} | +{a} -{d}" for p, a, d, _ in rows)
    used = len(stat)
    for path, a, d, body in sorted(rows, key=lambda r: len(r[3])):
        if previous and previous.get(path) == hashes[path]:
            unchanged.append(path)
            continue
        if used + len(body) <= budget:
            send.append((path, body))
            used += len(body)
    sent_paths = {p for p, _ in send}
    rest = [p for p, _, _, _ in rows if p not in sent_paths and p not in unchanged]
    cmd = f"git diff {base} -- <path>" if base else "git diff -- <path>"
    parts = [f"DIFFSTAT ({len(rows)} files)", stat]
    if untracked:
        parts += ["", "UNTRACKED FILES (new, not in git diff; read them directly): " + ", ".join(untracked[:60])]
    if unchanged:
        parts += ["", "Unchanged since your last review (not repeated): " + ", ".join(unchanged)]
    if rest:
        parts += ["", f"Not included to save tokens (run `{cmd}` for any you need): " + ", ".join(rest)]
    order = {p: i for i, (p, _, _, _) in enumerate(rows)}
    if send:
        parts += ["", "\n".join(body.rstrip("\n") for _, body in sorted(send, key=lambda x: order[x[0]]))]
    return "\n".join(parts), hashes


# ============================================================================ compaction
def compact_handoff(role: str, state: dict, contract: str, changed: list, open_findings: str = "", context_tokens: int = 0) -> str:
    """What a fresh session needs to carry on, built by Relay from its own records (no model call)."""
    plan = state.get("plan") or {}
    report = state.get("report") or {}
    lines = [f"ORCHESTRATOR · CONTEXT HANDOFF · your previous session for this role grew to about {context_tokens:,} tokens, so Relay "
             "started a fresh one to keep every later turn cheaper. The worktree holds all the work. This is the state; do not redo "
             "finished work and do not re-read the whole repository."]
    if plan.get("summary") or plan.get("plan"):
        lines += ["", "PLAN", str(plan.get("summary") or ""), str(plan.get("plan") or "")[:3000]]
    if contract:
        lines += ["", "ACCEPTANCE CONTRACT (current status)", contract]
    if open_findings:
        lines += ["", "OPEN BLOCKING FINDINGS", open_findings]
    lines += ["", f"PROGRESS · work package #{state.get('turn') or 0} · phase {state.get('phase') or '?'}"]
    if state.get("instruction_summary"):
        lines.append(f"Current package: {state['instruction_summary']}")
    if report.get("summary"):
        lines.append(f"Last worker report ({report.get('status', '?')}): {report['summary']}")
    if changed:
        lines.append("Changed files: " + ", ".join(f"{c.get('status', '')} {c.get('path', '')}".strip() for c in changed[:40])
                     + (f" (+{len(changed) - 40} more)" if len(changed) > 40 else ""))
    blocked = state.get("blocked_checks") or []
    if blocked:
        lines.append("Checks that could not run: " + "; ".join(f"{b.get('check')}: {b.get('reason')}" for b in blocked[:8]))
    return "\n".join(x for x in lines if x is not None)


# ============================================================================ report
def _day(ts: float) -> str:
    return datetime.fromtimestamp(ts, timezone.utc).strftime("%Y-%m-%d")


def report(tasks: list, days: int = 30, now_ts: float | None = None) -> dict:
    """Where tokens went across tasks: totals by role, agent and prompt section, cache hits, trend per day, top tasks."""
    now_ts = now_ts or time.time()
    since = now_ts - days * 86400
    zero = lambda: {"turns": 0, "input": 0, "cached": 0, "output": 0, "prompt": 0, "cost_usd": 0.0}  # noqa: E731
    by_role, by_agent, by_day, sections = {}, {}, {}, {}
    total = zero()
    rows = []
    measured_turns = 0
    for t in tasks:
        log = ((t.get("metrics") or {}).get("log") or [])
        trow = zero()
        tsec = {}
        for e in log:
            if float(e.get("end") or 0) < since:
                continue
            agent, role = e.get("agent") or "?", e.get("role") or "?"
            usage = {"input": e.get("input"), "cached": e.get("cached"), "output": e.get("output"), "cache_write": e.get("cache_write")}
            ptok = prompt_tokens(agent, usage)
            for bucket in (total, trow, by_role.setdefault(role, zero()), by_agent.setdefault(agent, zero()),
                           by_day.setdefault(_day(float(e.get("end") or now_ts)), zero())):
                bucket["turns"] += 1
                bucket["input"] += int(e.get("input") or 0)
                bucket["cached"] += int(e.get("cached") or 0)
                bucket["output"] += int(e.get("output") or 0)
                bucket["prompt"] += ptok
                bucket["cost_usd"] = round(bucket["cost_usd"] + float(e.get("cost_usd") or 0), 5)
            if e.get("sections"):
                measured_turns += 1
                for k, v in e["sections"].items():
                    sections[k] = sections.get(k, 0) + int(v or 0)
                    tsec[k] = tsec.get(k, 0) + int(v or 0)
        if trow["turns"]:
            rows.append({"id": t.get("id"), "number": t.get("number"), "name": t.get("name"), "status": t.get("status"), **trow,
                         "cache_ratio": round(trow["cached"] / trow["prompt"], 3) if trow["prompt"] else 0.0,
                         "sections": tsec, "updated_at": t.get("updated_at")})

    def finish(b):
        return {**b, "cache_ratio": round(b["cached"] / b["prompt"], 3) if b["prompt"] else 0.0,
                "output_share": round(b["output"] / (b["prompt"] + b["output"]), 3) if (b["prompt"] + b["output"]) else 0.0}

    ordered = {k: sections[k] for k in SECTION_ORDER if k in sections}
    ordered.update({k: v for k, v in sections.items() if k not in ordered})
    return {
        "days": days, "total": finish(total), "roles": {k: finish(v) for k, v in by_role.items()},
        "agents": {k: finish(v) for k, v in by_agent.items()},
        "sections": {k: {"chars": v, "tokens": est(v)} for k, v in ordered.items()}, "measured_turns": measured_turns,
        "trend": [{"day": d, **finish(v)} for d, v in sorted(by_day.items())],
        "tasks": sorted(rows, key=lambda r: r["cost_usd"] or r["prompt"], reverse=True)[:25],
    }
